geode-core/src/distributedTest/java/org/apache/geode/cache30/CacheStatisticsPartitionedRegionDUnitTest.java

```diff
@@ -0,0 +1,577 @@
+    region.get(key3); // miss on existing entry
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 2 == rStats.getMissCount());
+    assertEquals(1, rStats.getHitCount());
+    assertEquals(2, rStats.getMissCount());
+    assertEquals(0.33f, rStats.getHitRatio(), 0.01f);
+
+    rStats.resetCounts();
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    region.invalidate(key);
+    region.get(key);
+
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(1, rStats.getMissCount());
+  }
+
+  /**
+   * Tests that the {@linkplain CacheStatistics#getLastAccessedTime last access time} and
+   * {@link CacheStatistics#getLastModifiedTime last modified time} are updated appropriately for a
+   * partitioned region with only one server.
+   */
+  @Test
+  public void testTimeStats() throws CacheException, InterruptedException {
+    final long ESTAT_RES = 100; // the resolution, in ms, of entry stats
+    String name = this.getUniqueName();
+    Object key = "KEY";
+    Object key2 = "KEY2";
+    Object value = "VALUE";
+    long before;
+    long after;
+    long oldBefore;
+    long oldAfter;
+
+    final int MSECS = 1000;
+
+    AttributesFactory factory = new AttributesFactory();
+    // factory.setScope(Scope.LOCAL);
+    factory.setStatisticsEnabled(true);
+
+    Region region = createPartitionedRegion(name, factory.create());
+    CacheStatistics rStats = region.getStatistics();
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    // In order for the stats to have a value, an entry must be accessed first.
+    region.get(key);
+    after = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+
+    {
+      final long localBefore = before;
+      final long localAfter = after;
+      GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+          .until(() -> (localBefore <= rStats.getLastAccessedTime()
+              && localAfter >= rStats.getLastAccessedTime()));
+    }
+    assertInRange(before, after, rStats.getLastModifiedTime());
+
+    oldBefore = before;
+    oldAfter = after;
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    region.get(key);
+    after = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+
+    {
+      final long localBefore = before;
+      final long localAfter = after;
+      GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+          .until(() -> localBefore <= rStats.getLastAccessedTime()
+              && localAfter >= rStats.getLastAccessedTime());
+    }
+    assertInRange(oldBefore, oldAfter, rStats.getLastModifiedTime());
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
+    region.put(key, value);
```

Comment:
Something this test should do is puts and gets with keys that will be in different buckets. I don't remember the details on how to do this. We want multiple buckets to make sure that lastAccessTIme and lastModifiedTime get the max when different buckets have different values. I think the hit/miss test also needs to do this. In the time test you would want to pause so that the timestamps are different in multiple buckets. I don't remember the details on how to do this. It ends up having to do with how the key gets hashed by the PartitionedRegion. This method shows how this is done: org.apache.geode.internal.cache.PartitionedRegionHelper.getHashKey
I think an easy way it to use Integer instances for as keys on your region since the hashCode in Integer is the value of the int. Then you can just have keys with values like 1 and 2 and know they will be in different buckets.

Reply:
I agree. I have changed the type of the keys to Integer.
I think the pausing you mention for the time tests can be implemented by calling the waitForClockToChange method you suggested in a previous comment before the "before" variable is assigned.